# Определение источников транзакций

<b>Описание:</b>

Внутри BigQuery имеется публичный датасет ga_sessions с описанием визитов пользователей на сайт интернет-магазина. Необходимо сформировать выгрузку за период 2 недели и для каждой совершенной транзакции составить цепочку источников визитов пользователя. Также необходимо определить топ-10 самых популярных цепочек. 

<b>Цель:</b>
1. Для каждой транзакции определить цепочку касаний.
2. Выявить топ-10 самых популярных цепочек касания, приведших к транзакции.

## Импорт библиотек и предобработка данных

Импортируем необходимые библиотеки и устанавливаем настройки

In [1]:
import pandas as pd
import numpy as np
from datetime import datetime, timedelta
import re
import warnings

warnings.filterwarnings('ignore')
pd.set_option('max_colwidth', None)
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

Загружаем таблицу

Таблица сформирована SQL-запросом:

SELECT DISTINCT fullVisitorId,

       visitNumber,        
       visitId, 
       visitStartTime, 
       date, 
       trafficSource.source,
       trafficSource.medium,
       hits.transaction.transactionId
FROM `bigquery-public-data.google_analytics_sample.ga_sessions_*`,

UNNEST(hits) AS hits

WHERE _table_suffix BETWEEN '20170719' AND '20170801'

In [2]:
df = pd.read_csv('bq-results-20230205-230025-1675638046829.csv', 
                dtype={'fullVisitorId':str,
                       'visitNumber':int,
                       'visitId':str,
                       'visitStartTime':str,
                       'date':str,
                       'source':str,
                       'medium':str,
                       'transactionId':str})
df.head()

fullVisitorId  visitNumber     visitId visitStartTime      date  \
0  2265977830656365501            1  1500821038     1500821038  20170723   
1  6474095028710839893            2  1500875167     1500875167  20170723   
2  6443653150131206845            3  1500822280     1500822280  20170723   
3  9860085665958246812            2  1500831228     1500831228  20170723   
4  7700344873342427798            4  1500816531     1500816531  20170723   

   source   medium transactionId  
0  google  organic           NaN  
1  google  organic           NaN  
2  google  organic           NaN  
3  google  organic           NaN  
4  google  organic           NaN

In [3]:
print(f'Дубликатов в таблице: {df.duplicated().sum()}', end='\n\n')
df.info()

Дубликатов в таблице: 0

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33146 entries, 0 to 33145
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   fullVisitorId   33146 non-null  object
 1   visitNumber     33146 non-null  int32 
 2   visitId         33146 non-null  object
 3   visitStartTime  33146 non-null  object
 4   date            33146 non-null  object
 5   source          33146 non-null  object
 6   medium          33146 non-null  object
 7   transactionId   521 non-null    object
dtypes: int32(1), object(7)
memory usage: 1.9+ MB


Дубликатов в таблице нет (так как мы выгрузили DISTINCT значения). Приведем данные к нужным форматам.

In [4]:
df['visitStartTime'] = pd.to_datetime(df['visitStartTime'], unit='s')
df['date'] = pd.to_datetime((df['date']), format='%Y%m%d')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33146 entries, 0 to 33145
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   fullVisitorId   33146 non-null  object        
 1   visitNumber     33146 non-null  int32         
 2   visitId         33146 non-null  object        
 3   visitStartTime  33146 non-null  datetime64[ns]
 4   date            33146 non-null  datetime64[ns]
 5   source          33146 non-null  object        
 6   medium          33146 non-null  object        
 7   transactionId   521 non-null    object        
dtypes: datetime64[ns](2), int32(1), object(5)
memory usage: 1.9+ MB


In [5]:
df.head()

fullVisitorId  visitNumber     visitId      visitStartTime  \
0  2265977830656365501            1  1500821038 2017-07-23 14:43:58   
1  6474095028710839893            2  1500875167 2017-07-24 05:46:07   
2  6443653150131206845            3  1500822280 2017-07-23 15:04:40   
3  9860085665958246812            2  1500831228 2017-07-23 17:33:48   
4  7700344873342427798            4  1500816531 2017-07-23 13:28:51   

        date  source   medium transactionId  
0 2017-07-23  google  organic           NaN  
1 2017-07-23  google  organic           NaN  
2 2017-07-23  google  organic           NaN  
3 2017-07-23  google  organic           NaN  
4 2017-07-23  google  organic           NaN

Проверяем, что даты в таблице те, что мы хотели получить.

In [6]:
print(f'Минимальная дата visitStartTime: {df["visitStartTime"].min()}')
print(f'Минимальная дата visitStartTime: {df["visitStartTime"].max()}')
print(f'Минимальная дата date: {df["date"].min()}')
print(f'Минимальная дата date: {df["date"].max()}')

Минимальная дата visitStartTime: 2017-07-19 07:00:27
Минимальная дата visitStartTime: 2017-08-02 06:59:53
Минимальная дата date: 2017-07-19 00:00:00
Минимальная дата date: 2017-08-01 00:00:00


Данные предобработаны, приступаем к исследовательскому анализу.

## Исследовательский анализ и подготовка финального датасета 

### Траназакции

Проверяем, что нет дублирующихся транзакций.

In [7]:
print(f'Всего транзакций: {df["transactionId"].count()}')
print(f'Уникальных транзакций: {df["transactionId"].nunique()}')

Всего транзакций: 521
Уникальных транзакций: 521


Создадим столбец `is_transaction` с флагом, что в данной строке есть транзакция.

In [8]:
df['transactionId'] = df['transactionId'].astype(str)
df['is_transaction'] = df['transactionId'].apply(lambda x: 1 if re.search('.*\d+.*', x) else 0)
df['transactionId'] = df['transactionId'].replace('nan', np.nan)
df['is_transaction'].value_counts()

0    32625
1      521
Name: is_transaction, dtype: int64

### Визиты

Смотрим, какие по счету визиты есть в таблице.

In [9]:
df['visitNumber'].value_counts()

1      24195
2       3911
3       1588
4        888
5        553
6        395
7        281
8        214
9        154
10       123
11       102
12        94
13        62
14        58
15        53
16        47
17        38
18        32
20        27
19        27
23        16
25        14
22        14
21        14
26        12
24        12
31        10
32        10
27         9
28         9
29         8
30         7
33         6
58         4
35         4
59         4
61         4
34         4
36         4
37         4
62         4
63         4
64         3
60         3
41         3
57         3
127        2
71         2
65         2
40         2
150        2
45         2
39         2
48         2
149        2
152        2
151        2
43         2
128        2
104        2
153        2
77         2
42         2
174        1
84         1
83         1
108        1
147        1
109        1
322        1
82         1
148        1
134        1
323        1
114        1
115        1
158        1

Необходимо в финальном датасете оставить только пользователей, совершивших транзакцию. А также только пользователей, для которых можно составить цепочку с первого визита. 

То есть пользователи, которые ранее 19.07.2017 совершали визиты, не попали к нам в таблицу, и корректную цепочку для них мы составить не можем.

In [10]:
good_visitors = df.query('is_transaction == 1')['fullVisitorId'].unique()
print(f'Всего пользователей, совершивших транзакцию: {len(good_visitors)}')

Всего пользователей, совершивших транзакцию: 492


In [11]:
print(f'Длина таблицы до: {len(df)}')
df_good = df.query('fullVisitorId in @good_visitors')
print(f'Длина таблицы после: {len(df_good)}')

Длина таблицы до: 33146
Длина таблицы после: 1730


Определяем минимальный порядковый номер визита.

In [12]:
min_visit = df_good.groupby('fullVisitorId').agg({'visitNumber':'min'}).reset_index()
min_visit

fullVisitorId  visitNumber
0    0010295111715775250            2
1    0024932550342595467            2
2    0047078955120420928            1
3    0080479763428955064            3
4    0082806901961150595            1
5    0097371986665596420           11
6    0126612099954375841            4
7    0143782636543876435            3
8    0169854884741652554            4
9    0188308435688462855            4
10    024507252193437459            1
11   0249659095832018179            1
12   0280589015308186459            3
13   0316080058721038988            1
14   0318476411656972479            1
15   0326671467950015369            2
16   0345672699449577691            4
17   0349598333236458498            1
18     03717831825503026            1
19   0377506370870345639            2
20   0389488682405584834            1
21   0396718069117936487            1
22   0500460722094613747            1
23   0509972280802528263            1
24   0535797652559788641            2
25   0546450430035162938            1
26   0548744394505098293            1
27   0572135527461851083            1
28    057693500927581077            1
29   0604314853571328266            1
30   0608915197735218105           16
31   0640040708763897460            2
32   0640433429877929929            3
33   0652202420147185334            1
34   0668741043882663600            2
35   0696402585067597025            2
36   0702289726935404752            2
37   0743501820485430157            1
38   0779199793341944056            1
39   0780253600713375371           10
40   0795389678082168807            3
41    081281456829886913            2
42   0815047945706399620            3
43   0842088345978971117            1
44   0843143557737586340            1
45    085319952392271706            1
46   0853682097133967435            1
47   0951967559171152344            1
48   0953889584295924240            2
49   0956108016718766611            1
50   0975931442450346746            1
51   1005326553546138067            8
52   1051896220301954151            6
53   1053771337658446740            1
54   1063109698143249629            1
55   1081954083914922307            1
56   1140574847502258865            3
57   1152695544079009829            1
58   1160204963389257110            7
59   1164110348121511511            1
60   1174647913878447921            1
61   1182675562893088634            1
62   1190814633327628141            4
63   1247282406356396417            1
64   1255368394012652243            2
65   1288317908843556558            1
66   1311003026247678016            1
67   1311682260438379600            1
68   1392760476578326953            1
69   1393633173328801040            2
70   1413421667450087159            1
71   1424355572088327868            1
72   1447257816279888170            1
73   1452613891025216774           12
74   1459959210477445011            1
75   1468478016521432136            1
76   1476938238176225678           13
77     15618322761092379            4
78   1563065767696461622            3
79    157019484210987302           14
80   1573919593896895308            2
81    160370198171303042            1
82    165109760635455356            1
83   1673188218512892503            1
84   1675171647898452800            5
85   1675608109098320028            2
86   1681975491896946073            1
87   1685335291483410828            1
88   1712066703099487652            8
89   1727715375435944327            2
90   1741172694620038876            1
91   1751309915031023278            1
92   1761084588077503545            1
93   1762540039573649288            2
94   1835132804289989526            8
95    186231215995844689            1
96   1863054289029276927            3
97   1875268638592946287            1
98    190492934318255272            1
99   1921325710867532207            3
100  1922510765164338251            1
101  1931723981901564088            1
102   195434588004709231            1
103  1958518504412290611            1
104  1996186587769697

In [13]:
good_visitors_2 = min_visit.query('visitNumber == 1')['fullVisitorId'].unique()
print('Всего пользователей, совершивших транзакцию и о которых имеются данные с первого захода: {}'
     .format(len(good_visitors_2)))

Всего пользователей, совершивших транзакцию и о которых имеются данные с первого захода: 286


В финальном датафрейме `df_fin` оставляем только пользователей, для которых есть данные с первого захода и которые совершили транзакцию.

In [14]:
df_fin = df_good.query('fullVisitorId in @good_visitors_2')
print(f'Длина итоговой таблицы для анализа: {len(df_fin)}', end='\n\n')
print(f'Количество транзакций в таблице: {df_fin["transactionId"].count()}')
print(f'Уникальных пользователей в таблице: {df_fin["fullVisitorId"].nunique()}')
df_fin.head()

Длина итоговой таблицы для анализа: 966

Количество транзакций в таблице: 297
Уникальных пользователей в таблице: 286


fullVisitorId  visitNumber     visitId      visitStartTime  \
9     9973665079624172058            1  1500870192 2017-07-24 04:23:12   
574   3563510904717248245            3  1500853402 2017-07-23 23:43:22   
1381   190492934318255272            6  1500860330 2017-07-24 01:38:50   
1382  5322397769480961983            4  1500857903 2017-07-24 00:58:23   
1546   599110551792149234            4  1500851005 2017-07-23 23:03:25   

           date    source   medium transactionId  is_transaction  
9    2017-07-23  (direct)   (none)           NaN               0  
574  2017-07-23    google  organic           NaN               0  
1381 2017-07-23    google  organic           NaN               0  
1382 2017-07-23    google  organic           NaN               0  
1546 2017-07-23  (direct)   (none)           NaN               0

### Источники

In [15]:
df_fin['source'].value_counts()

(direct)            542
google              405
sites.google.com     10
mail.google.com       5
youtube.com           4
Name: source, dtype: int64

Удалим окончание `.com` из названия источника, а также скобки при прямом заходе.

In [16]:
df_fin['source'] = df_fin['source'].apply(lambda x:  re.sub('(\.com$)', '', x))
df_fin['source'] = df_fin['source'].replace('(direct)', 'direct')
df_fin['source'].value_counts()

direct          542
google          405
sites.google     10
mail.google       5
youtube           4
Name: source, dtype: int64

In [17]:
df_fin['medium'].value_counts()

(none)      542
organic     329
cpc          60
referral     19
cpm          16
Name: medium, dtype: int64

Тоже удалим скобки для прямых заходов.

In [18]:
df_fin['medium'] = df_fin['medium'].replace('(none)', 'none')
df_fin['medium'].value_counts()

none        542
organic     329
cpc          60
referral     19
cpm          16
Name: medium, dtype: int64

Добавим колонку с объединенными source и medium составляющими источника.

In [19]:
df_fin['source_medium'] = df_fin['source'] + '_' + df_fin['medium']
df_fin['source_medium'].value_counts()

direct_none              542
google_organic           329
google_cpc                60
google_cpm                16
sites.google_referral     10
mail.google_referral       5
youtube_referral           4
Name: source_medium, dtype: int64

Видим, что чаще всего источником визита был прямой заход, а не по размеченным источникам.

In [20]:
df_fin = df_fin.sort_values(['fullVisitorId', 'visitStartTime'])
df_fin.head()

fullVisitorId  visitNumber     visitId      visitStartTime  \
33023  0047078955120420928            1  1501541991 2017-07-31 22:59:51   
33024  0047078955120420928            1  1501541991 2017-07-31 22:59:51   
25246  0082806901961150595            1  1501008172 2017-07-25 18:42:52   
25247  0082806901961150595            1  1501008172 2017-07-25 18:42:52   
11995   024507252193437459            1  1501250419 2017-07-28 14:00:19   

            date  source medium    transactionId  is_transaction source_medium  
33023 2017-07-31  direct   none              NaN               0   direct_none  
33024 2017-07-31  direct   none  ORD201707312556               1   direct_none  
25246 2017-07-25  direct   none              NaN               0   direct_none  
25247 2017-07-25  direct   none  ORD201707252631               1   direct_none  
11995 2017-07-28  direct   none              NaN               0   direct_none

##### Вывод:
- Подготовлен финальный датасет для анализа источников транзакций
- В нем 966 строк и 297 транзакций 
- 286 уникальных пользователей, то есть имеются пользователи, совершившие не одну транзакцию 
- Самым популярным источником визита является прямой заход.

## Определение цепочек для транзакций

Составим таблицу, где для каждого пользователя, визита и транзакции будет указан источник без дубликатов.

In [21]:
# уникальные номера визитов для пользователя
visit_source = df_fin[['fullVisitorId', 'visitNumber', 'source_medium']].drop_duplicates()

# уникальные сочетания номера пользователя, номера визита, номера транзакции
visit_transactions = df_fin[['fullVisitorId', 'visitNumber',
        'transactionId', 'is_transaction']].drop_duplicates()

# оставляем для каждого номера визита с транзакцией только запись, где была транзакция 
# (т.к. источник для каждого визита единственный)
visit_transactions = visit_transactions.query('is_transaction == 1')

# и соединяем таблицы
df_transactions = visit_source.merge(visit_transactions, 
                                     on=['fullVisitorId', 'visitNumber'], 
                                     how='left')

df_transactions['is_transaction'] = df_transactions['is_transaction'].fillna(0).astype(int)
df_transactions.head()

fullVisitorId  visitNumber         source_medium    transactionId  \
0  0047078955120420928            1           direct_none  ORD201707312556   
1  0082806901961150595            1           direct_none  ORD201707252631   
2   024507252193437459            1           direct_none              NaN   
3   024507252193437459            2           direct_none              NaN   
4   024507252193437459            3  mail.google_referral  ORD201708012322   

   is_transaction  
0               1  
1               1  
2               0  
3               0  
4               1

Составим списки источников, транзакций для каждого пользователя.

In [22]:
df_grouped_transactions = (df_transactions
                           .groupby('fullVisitorId')
                           .agg({'source_medium': lambda x: list(x), 
                                 'transactionId': lambda x: list(x),
                                 'is_transaction': lambda x: list(x)}))

df_grouped_transactions.head()

source_medium  \
fullVisitorId                                                           
0047078955120420928                                     [direct_none]   
0082806901961150595                                     [direct_none]   
024507252193437459   [direct_none, direct_none, mail.google_referral]   
0249659095832018179                                     [direct_none]   
0316080058721038988      [google_cpm, google_organic, google_organic]   

                                   transactionId is_transaction  
fullVisitorId                                                    
0047078955120420928            [ORD201707312556]            [1]  
0082806901961150595            [ORD201707252631]            [1]  
024507252193437459   [nan, nan, ORD201708012322]      [0, 0, 1]  
0249659095832018179            [ORD201707242363]            [1]  
0316080058721038988  [nan, ORD201707192364, nan]      [0, 1, 0]

Добавим поле `N_of_visit` с индексом транзакции

In [23]:
df_grouped_transactions['N_of_visit'] = (df_grouped_transactions['is_transaction']
                                         .apply(lambda x: np.where(np.array(x) == 1)[0]))

df_grouped_transactions['N_of_visit'].value_counts()

[0]             176
[1]              55
[2]              22
[3]              13
[4]               5
[5]               3
[7]               3
[0, 1]            1
[1, 7]            1
[0, 1, 2, 3]      1
[1, 2, 4]         1
[2, 3]            1
[0, 1]            1
[16]              1
[0, 1]            1
[4, 5]            1
Name: N_of_visit, dtype: int64

Чаще всего транзация совершалась в первый визит (индекс=0).

In [24]:
visits_rating = df_fin.query('is_transaction == 1')['visitNumber'].value_counts().to_frame()
visits_rating['%_of_all'] = round(visits_rating['visitNumber'] * 100 / visits_rating['visitNumber'].sum(), 1)
visits_rating

visitNumber  %_of_all
1           180      60.6
2            61      20.5
3            24       8.1
4            14       4.7
5             7       2.4
6             6       2.0
8             4       1.3
17            1       0.3

Добавим поля для дальнейшей обработки.

In [25]:
# Список цепочек источников для каждой транзакции (если у пользователя было 2 транзакции, то
# в списке будет 2 цепочки)
df_grouped_transactions['list_source'] = (df_grouped_transactions
                                          .apply(lambda x: 
        [x['source_medium'][:x['N_of_visit'][i] + 1] for i in range(len(x['N_of_visit']))], axis=1))

# Список транзакций
df_grouped_transactions['list_transactions'] = (df_grouped_transactions
                                                .apply(lambda x: 
        [x['transactionId'][x['N_of_visit'][i]] for i in range(len(x['N_of_visit']))], axis=1))

# Список из идущих подряд уникальных источников и без прямых заходов
df_grouped_transactions['source_short'] = (df_grouped_transactions
                                         .apply(lambda x: 
        [list(filter(lambda y: y != 'direct_none', 
                     [x['list_source'][j][i] for i in range(len(x['list_source'][j])) 
          if i == 0 or (i > 0 and x['list_source'][j][i] != x['list_source'][j][i-1])
         ]))
         for j in range(len(x['N_of_visit']))
         
        ], axis=1))
                                         
df_grouped_transactions.head()                            

source_medium  \
fullVisitorId                                                           
0047078955120420928                                     [direct_none]   
0082806901961150595                                     [direct_none]   
024507252193437459   [direct_none, direct_none, mail.google_referral]   
0249659095832018179                                     [direct_none]   
0316080058721038988      [google_cpm, google_organic, google_organic]   

                                   transactionId is_transaction N_of_visit  \
fullVisitorId                                                                
0047078955120420928            [ORD201707312556]            [1]        [0]   
0082806901961150595            [ORD201707252631]            [1]        [0]   
024507252193437459   [nan, nan, ORD201708012322]      [0, 0, 1]        [2]   
0249659095832018179            [ORD201707242363]            [1]        [0]   
0316080058721038988  [nan, ORD201707192364, nan]      [0, 1, 0]        [1]   

                                                            list_source  \
fullVisitorId                                                             
0047078955120420928                                     [[direct_none]]   
0082806901961150595                                     [[direct_none]]   
024507252193437459   [[direct_none, direct_none, mail.google_referral]]   
0249659095832018179                                     [[direct_none]]   
0316080058721038988                      [[google_cpm, google_organic]]   

                     list_transactions                    source_short  
fullVisitorId                                                           
0047078955120420928  [ORD201707312556]                            [[]]  
0082806901961150595  [ORD201707252631]                            [[]]  
024507252193437459   [ORD201708012322]        [[mail.google_referral]]  
0249659095832018179  [ORD201707242363]                            [[]]  
0316080058721038988  [ORD201707192364]  [[google_cpm, google_organic]]

In [26]:
df_grouped_transactions.query('list_transactions.str.len() > 1').head(2)

source_medium  \
fullVisitorId                                                                                                                                                                                                                                                                                                                                           
4042880551685271731                                                                                                                                                                                                                                                                                                        [direct_none, direct_none]   
4376572204985925602  [google_organic, google_organic, google_organic, google_organic, google_organic, google_organic, google_organic, google_organic, google_organic, google_organic, google_organic, google_organic, google_organic, google_organic, google_organic, google_organic, google_organic, google_organic, google_organic, google_organic]   

                                                                                                                                    transactionId  \
fullVisitorId                                                                                                                                       
4042880551685271731                                                                                            [ORD201707202545, ORD201707202429]   
4376572204985925602  [nan, ORD201707192493, nan, nan, nan, nan, nan, ORD201707192400, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan]   

                                                                   is_transaction  \
fullVisitorId                                                                       
4042880551685271731                                                        [1, 1]   
4376572204985925602  [0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]   

                    N_of_visit  \
fullVisitorId                    
4042880551685271731     [0, 1]   
4376572204985925602     [1, 7]   

                                                                                                                                                                              list_source  \
fullVisitorId                                                                                                                                                                               
4042880551685271731                                                                                                                           [[direct_none], [direct_none, direct_none]]   
4376572204985925602  [[google_organic, google_organic], [google_organic, google_organic, google_organic, google_organic, google_organic, google_organic, google_organic, google_organic]]   

                                      list_transactions  \
fullVisitorId                                             
4042880551685271731  [ORD201707202545, ORD201707202429]   
4376572204985925602  [ORD201707192493, ORD201707192400]   

                                             source_short  
fullVisitorId                                              
4042880551685271731                              [[], []]  
4376572204985925602  [[google_organic], [google_organic]]

Например, у пользователей `4042880551685271731` и `4376572204985925602` было по 2 транзакции. У первого пользователя транзакция была в первый и второй визит, у второго - во второй и восьмой.

В колонке `list_transactions` перечислены номера транзакций. 

В колонке `list_source` для каждой транзакции список источников захода. Например, для второй транзакции второго пользователя считаем цепочку источников с самого первого визита, и у него будет цепочка из 8 касаний.

Поскольку в итоге нас интересуют размеченные заходы и топ цепочек, будем считать, что несколько переходов подряд с одного источника можно сложить в один переход (неважно, сколько раз подряд пользователь перешел с одного и того же источника). Поэтому в колонке `source_short` оставлены только уникальные источники для каждой транзакции (если первая и вторая одинаковые - считаем за один источник) и удалены прямые заходы. 

In [27]:
list_transactions = df_grouped_transactions['list_transactions'].to_list()
list_sources = df_grouped_transactions['list_source'].to_list()

transactions_tuple = [(list_transactions[j][i], ' => '.join(list_sources[j][i])) 
                      for j in range(len(list_transactions)) for i in range(len(list_transactions[j]))]

transactions_dict = dict(transactions_tuple)
transactions_dict

{'ORD201707312556': 'direct_none',
 'ORD201707252631': 'direct_none',
 'ORD201708012322': 'direct_none => direct_none => mail.google_referral',
 'ORD201707242363': 'direct_none',
 'ORD201707192364': 'google_cpm => google_organic',
 'ORD201707282356': 'direct_none',
 'ORD201707282299': 'google_organic',
 'ORD201708012493': 'direct_none',
 'ORD201707312543': 'direct_none',
 'ORD201707221572': 'direct_none',
 'ORD201707212233': 'direct_none => direct_none => direct_none => direct_none => direct_none => direct_none',
 'ORD201708011814': 'google_organic => google_organic => direct_none => direct_none',
 'ORD201707231817': 'direct_none',
 'ORD201707252455': 'direct_none',
 'ORD201707231818': 'direct_none',
 'ORD201707221645': 'direct_none => direct_none => direct_none => direct_none',
 'ORD201707262668': 'google_organic',
 'ORD201707252458': 'google_organic => google_organic',
 'ORD201707252624': 'google_cpm',
 'ORD201707272410': 'direct_none => direct_none => direct_none => direct_none',
 '

Итоговая таблица `transactions_chain` с цепочкой касаний для каждой транзакции.

In [28]:
transactions_chain = pd.DataFrame.from_dict(transactions_dict, orient='index').reset_index()
transactions_chain.columns = ['transactionId', 'chain']
transactions_chain

transactionId  \
0    ORD201707312556   
1    ORD201707252631   
2    ORD201708012322   
3    ORD201707242363   
4    ORD201707192364   
5    ORD201707282356   
6    ORD201707282299   
7    ORD201708012493   
8    ORD201707312543   
9    ORD201707221572   
10   ORD201707212233   
11   ORD201708011814   
12   ORD201707231817   
13   ORD201707252455   
14   ORD201707231818   
15   ORD201707221645   
16   ORD201707262668   
17   ORD201707252458   
18   ORD201707252624   
19   ORD201707272410   
20   ORD201707312635   
21   ORD201708012538   
22   ORD201707282407   
23   ORD201707282413   
24   ORD201707301587   
25   ORD201707231841   
26   ORD201707212220   
27   ORD201707282340   
28   ORD201707242280   
29   ORD201707212301   
30   ORD201707252313   
31   ORD201707252628   
32   ORD201707272475   
33   ORD201707252392   
34   ORD201707262709   
35   ORD201707272478   
36   ORD201708012355   
37   ORD201707242451   
38   ORD201707202652   
39   ORD201707192738   
40   ORD201707252308   
41   ORD201707272495   
42   ORD201707282403   
43   ORD201707272162   
44   ORD201707212395   
45   ORD201707252555   
46   ORD201707301701   
47   ORD201707312442   
48   ORD201708012371   
49   ORD201707291534   
50   ORD201707262624   
51   ORD201707262697   
52   ORD201707311786   
53   ORD201707212084   
54   ORD201707202321   
55   ORD201707202606   
56   ORD201707231848   
57   ORD201707192465   
58   ORD201707212208   
59   ORD201707192504   
60   ORD201707192486   
61   ORD201707252407   
62   ORD201707282300   
63   ORD201707282309   
64   ORD201707301640   
65   ORD201707291465   
66   ORD201707262646   
67   ORD201707192281   
68   ORD201708012497   
69   ORD201707192357   
70   ORD201707312647   
71   ORD201707301721   
72   ORD201707291489   
73   ORD201707252479   
74   ORD201707262398   
75   ORD201707312289   
76   ORD201707242459   
77   ORD201707272506   
78   ORD201707242351   
79   ORD201707301532   
80   ORD201707272515   
81   ORD201707242223   
82   ORD201707242442   
83   ORD201707282399   
84   ORD201708012576   
85   ORD201707192472   
86   ORD201707202570   
87   ORD201707312613   
88   ORD201707262740   
89   ORD201707312390   
90   ORD201707212356   
91   ORD201707221573   
92   ORD201707282402   
93   ORD201707242495   
94   ORD201707192458   
95   ORD201707252521   
96   ORD201707212241   
97   ORD201707272378   
98   ORD201708012479   
99   ORD201707312284   
100  ORD201707231538   
101  ORD201707262669   
102  ORD201707221487   
103  ORD201707212302   
104  ORD201707272497   
105  ORD201707262718   
106  ORD201707301681   
107  ORD201707272528   
108  ORD201707221593   
109  ORD201707291494   
110  ORD201707291524   
111  ORD201707262691   
112  ORD201707312584   
113  ORD201707202545   
114  ORD201707202429   
115  ORD201707212374   
116  ORD201707312423   
117  ORD201707242382   
118  ORD201707252565   
119  ORD201707282359   
120  ORD201707312558   
121  ORD201707272407   
122  ORD201707192493   
123  ORD201707192400   
124  ORD201707312302   
125  ORD201707262696   
126  ORD201708012305   
127  ORD201707312594   
128  ORD201707192499   
129  ORD201707242355   
130  ORD201707272424   
131  ORD201707282381   
132  ORD201707272309   
133  ORD201707272361   
134  ORD201707312279   
135  ORD201707291482   
136  ORD201707312608   
137  ORD201707202597   
138  ORD201707231907   
139  ORD201707312536   
140  ORD201707202664   
141  ORD201707231694   
142  ORD201707262705   
143  ORD201707202523   
144  ORD201708012443   
145  ORD201707242431   
146  ORD201707262722   
147  ORD201707282322   
148  ORD201707252506   
149  ORD201707312560   
150  ORD201707202397   
151  ORD201707301706   
152  ORD201707192413   
153  ORD201707312650   
154  ORD201707282346   
155  ORD201707242371   
156  ORD201707272423   
157  ORD201707282224   
158  ORD201707282373   
159  ORD201707282231   
160  ORD201707291511   
161  ORD201707262685   
162  ORD201707312271   
163  ORD201707312627   
164  ORD201707212351   
165  ORD201707212405   

In [29]:
list_unique_sources = df_grouped_transactions['source_short'].to_list()

transactions_short_tuple = [(list_transactions[j][i], ' => '.join(list_unique_sources[j][i])) 
                          for j in range(len(list_transactions)) for i in range(len(list_transactions[j])) 
                          if list_unique_sources[j][i]]
transactions_short_dict = dict(transactions_short_tuple)
transactions_short_dict

{'ORD201708012322': 'mail.google_referral',
 'ORD201707192364': 'google_cpm => google_organic',
 'ORD201707282299': 'google_organic',
 'ORD201708011814': 'google_organic',
 'ORD201707262668': 'google_organic',
 'ORD201707252458': 'google_organic',
 'ORD201707252624': 'google_cpm',
 'ORD201707212220': 'google_cpc',
 'ORD201707242280': 'google_organic',
 'ORD201707212301': 'google_organic',
 'ORD201707252313': 'google_organic',
 'ORD201707272478': 'google_organic',
 'ORD201707242451': 'google_organic',
 'ORD201707192738': 'youtube_referral',
 'ORD201707252308': 'google_organic => google_cpc => google_organic',
 'ORD201707282403': 'google_organic',
 'ORD201707212395': 'google_organic',
 'ORD201707252555': 'google_organic',
 'ORD201707312442': 'google_organic',
 'ORD201707291534': 'google_organic',
 'ORD201707262624': 'google_organic',
 'ORD201707311786': 'google_organic',
 'ORD201707212084': 'google_organic',
 'ORD201707202321': 'google_cpm => google_organic',
 'ORD201707192465': 'google_

Таблица `transactions_chain_short` с уникальными идущими подряд источниками перехода без прямых заходов

In [30]:
transactions_chain_short = pd.DataFrame.from_dict(transactions_short_dict, orient='index').reset_index()
transactions_chain_short.columns = ['transactionId', 'chain']
transactions_chain_short

transactionId                                           chain
0    ORD201708012322                            mail.google_referral
1    ORD201707192364                    google_cpm => google_organic
2    ORD201707282299                                  google_organic
3    ORD201708011814                                  google_organic
4    ORD201707262668                                  google_organic
5    ORD201707252458                                  google_organic
6    ORD201707252624                                      google_cpm
7    ORD201707212220                                      google_cpc
8    ORD201707242280                                  google_organic
9    ORD201707212301                                  google_organic
10   ORD201707252313                                  google_organic
11   ORD201707272478                                  google_organic
12   ORD201707242451                                  google_organic
13   ORD201707192738                                youtube_referral
14   ORD201707252308  google_organic => google_cpc => google_organic
15   ORD201707282403                                  google_organic
16   ORD201707212395                                  google_organic
17   ORD201707252555                                  google_organic
18   ORD201707312442                                  google_organic
19   ORD201707291534                                  google_organic
20   ORD201707262624                                  google_organic
21   ORD201707311786                                  google_organic
22   ORD201707212084                                  google_organic
23   ORD201707202321                    google_cpm => google_organic
24   ORD201707192465                                  google_organic
25   ORD201707192504                            mail.google_referral
26   ORD201707252407                                  google_organic
27   ORD201707282309                                      google_cpm
28   ORD201707301640                                  google_organic
29   ORD201707262646                           sites.google_referral
30   ORD201707192281                                      google_cpc
31   ORD201707301721                                  google_organic
32   ORD201707291489                                  google_organic
33   ORD201707262398                                  google_organic
34   ORD201707242351                                  google_organic
35   ORD201707242223                                  google_organic
36   ORD201707212356                                  google_organic
37   ORD201707242495                                  google_organic
38   ORD201707192458                                      google_cpc
39   ORD201707252521                                  google_organic
40   ORD201707272378                                  google_organic
41   ORD201707272497                                  google_organic
42   ORD201707262718                                      google_cpc
43   ORD201707301681                                  google_organic
44   ORD201707272528                    google_cpc => google_organic
45   ORD201707291494                                  google_organic
46   ORD201707291524                                  google_organic
47   ORD201707212374                                  google_organic
48   ORD201707242382                                  google_organic
49   ORD201707282359                                  google_organic
50   ORD201707272407                                  google_organic
51   ORD201707192493                                  google_organic
52   ORD201707192400                                  google_organic
53   ORD201707192499                                      google_cpc
54   ORD201707291482                                  google_organic
55   ORD201707312608                                  google_organic
56   ORD201707202664                                      google_cpc
57   

Таблица `top_chain` с самыми популярными цепочками перехода для совершения транзакций. Топ-10 не набралось, потому что исследовался небольшой период.

In [31]:
top_chain = (transactions_chain_short['chain']
             .value_counts()
             .to_frame()
             .rename(columns={'chain':'n_chains'}))
top_chain

n_chains
google_organic                                        95
google_cpc                                            18
google_cpm                                             7
google_cpc => google_organic                           4
mail.google_referral                                   2
google_cpm => google_organic                           2
sites.google_referral                                  2
youtube_referral                                       1
google_organic => google_cpc => google_organic         1

## Заключение

Сформированы результирующие таблицы - таблица с цепочкой касаний для каждой транзакции `transactions_chain` и таблица с самыми популялярными цепочками касаний - `top_chain`.

Поскольку анализируемый период небольшой, получилось всего 9 уникальных цепочек (без прямых заходов).

In [32]:
transactions_chain

transactionId  \
0    ORD201707312556   
1    ORD201707252631   
2    ORD201708012322   
3    ORD201707242363   
4    ORD201707192364   
5    ORD201707282356   
6    ORD201707282299   
7    ORD201708012493   
8    ORD201707312543   
9    ORD201707221572   
10   ORD201707212233   
11   ORD201708011814   
12   ORD201707231817   
13   ORD201707252455   
14   ORD201707231818   
15   ORD201707221645   
16   ORD201707262668   
17   ORD201707252458   
18   ORD201707252624   
19   ORD201707272410   
20   ORD201707312635   
21   ORD201708012538   
22   ORD201707282407   
23   ORD201707282413   
24   ORD201707301587   
25   ORD201707231841   
26   ORD201707212220   
27   ORD201707282340   
28   ORD201707242280   
29   ORD201707212301   
30   ORD201707252313   
31   ORD201707252628   
32   ORD201707272475   
33   ORD201707252392   
34   ORD201707262709   
35   ORD201707272478   
36   ORD201708012355   
37   ORD201707242451   
38   ORD201707202652   
39   ORD201707192738   
40   ORD201707252308   
41   ORD201707272495   
42   ORD201707282403   
43   ORD201707272162   
44   ORD201707212395   
45   ORD201707252555   
46   ORD201707301701   
47   ORD201707312442   
48   ORD201708012371   
49   ORD201707291534   
50   ORD201707262624   
51   ORD201707262697   
52   ORD201707311786   
53   ORD201707212084   
54   ORD201707202321   
55   ORD201707202606   
56   ORD201707231848   
57   ORD201707192465   
58   ORD201707212208   
59   ORD201707192504   
60   ORD201707192486   
61   ORD201707252407   
62   ORD201707282300   
63   ORD201707282309   
64   ORD201707301640   
65   ORD201707291465   
66   ORD201707262646   
67   ORD201707192281   
68   ORD201708012497   
69   ORD201707192357   
70   ORD201707312647   
71   ORD201707301721   
72   ORD201707291489   
73   ORD201707252479   
74   ORD201707262398   
75   ORD201707312289   
76   ORD201707242459   
77   ORD201707272506   
78   ORD201707242351   
79   ORD201707301532   
80   ORD201707272515   
81   ORD201707242223   
82   ORD201707242442   
83   ORD201707282399   
84   ORD201708012576   
85   ORD201707192472   
86   ORD201707202570   
87   ORD201707312613   
88   ORD201707262740   
89   ORD201707312390   
90   ORD201707212356   
91   ORD201707221573   
92   ORD201707282402   
93   ORD201707242495   
94   ORD201707192458   
95   ORD201707252521   
96   ORD201707212241   
97   ORD201707272378   
98   ORD201708012479   
99   ORD201707312284   
100  ORD201707231538   
101  ORD201707262669   
102  ORD201707221487   
103  ORD201707212302   
104  ORD201707272497   
105  ORD201707262718   
106  ORD201707301681   
107  ORD201707272528   
108  ORD201707221593   
109  ORD201707291494   
110  ORD201707291524   
111  ORD201707262691   
112  ORD201707312584   
113  ORD201707202545   
114  ORD201707202429   
115  ORD201707212374   
116  ORD201707312423   
117  ORD201707242382   
118  ORD201707252565   
119  ORD201707282359   
120  ORD201707312558   
121  ORD201707272407   
122  ORD201707192493   
123  ORD201707192400   
124  ORD201707312302   
125  ORD201707262696   
126  ORD201708012305   
127  ORD201707312594   
128  ORD201707192499   
129  ORD201707242355   
130  ORD201707272424   
131  ORD201707282381   
132  ORD201707272309   
133  ORD201707272361   
134  ORD201707312279   
135  ORD201707291482   
136  ORD201707312608   
137  ORD201707202597   
138  ORD201707231907   
139  ORD201707312536   
140  ORD201707202664   
141  ORD201707231694   
142  ORD201707262705   
143  ORD201707202523   
144  ORD201708012443   
145  ORD201707242431   
146  ORD201707262722   
147  ORD201707282322   
148  ORD201707252506   
149  ORD201707312560   
150  ORD201707202397   
151  ORD201707301706   
152  ORD201707192413   
153  ORD201707312650   
154  ORD201707282346   
155  ORD201707242371   
156  ORD201707272423   
157  ORD201707282224   
158  ORD201707282373   
159  ORD201707282231   
160  ORD201707291511   
161  ORD201707262685   
162  ORD201707312271   
163  ORD201707312627   
164  ORD201707212351   
165  ORD201707212405   

In [33]:
top_chain

n_chains
google_organic                                        95
google_cpc                                            18
google_cpm                                             7
google_cpc => google_organic                           4
mail.google_referral                                   2
google_cpm => google_organic                           2
sites.google_referral                                  2
youtube_referral                                       1
google_organic => google_cpc => google_organic         1

##### Вывод
Таким образом, наиболее популярным рекламным источником является google - именно для него и его сервисов наиболее часто встречаются переходы, приведшие к транзакциям.

Также отметим, что чаще всего транзакция совершалась в первый визит - в 60,6% случаев, в 20,5% - во второй визит и в 8,1% случаев - в 3-й визит.